### ISSUE TITLE
`fix` doesn't settle on the first time

### ISSUE THREAD

**Issue by juhoautio** [sqlfluff/sqlfluff#2584]
This is based on a previous comment in another issue that was closed: https://github.com/sqlfluff/sqlfluff/issues/2134#issuecomment-1030589266

## Expected Behaviour

Running `sqlfluff fix` modifies query format – but running it again on the same file should not make any further changes.

## Observed Behaviour

Summary: The result changes considerably also on the 2nd run. After that fix has settled, ie. the format doesn't change any more if `fix` is run again.

## Steps to Reproduce

`my_query.sql`:
```sql
SELECT
    abs(round(a.metricx-b.metricx)) as col_c_rel_diff,
    abs((round(a.metricx-b.metricx)/a.metricx)*100) as metric_x_rel_diff
FROM foo_bar_report a
LEFT JOIN xxx_yyy_report b
ON a.event_date = b.event_date;
```

Execute this once:
```
sqlfluff fix my_query.sql -f
```
It exits with 0.

Output:
```
==== finding fixable violations ====
== [my_query.sql] FAIL
L:   2 | P:  24 | L006 | Missing whitespace before -
L:   2 | P:  24 | L006 | Missing whitespace after -
L:   2 | P:  37 | L010 | Keywords must be consistently upper case.
L:   3 | P:  25 | L006 | Missing whitespace before -
L:   3 | P:  25 | L006 | Missing whitespace after -
L:   3 | P:  36 | L006 | Missing whitespace before /
L:   3 | P:  36 | L006 | Missing whitespace after /
L:   3 | P:  47 | L006 | Missing whitespace before *
L:   3 | P:  47 | L006 | Missing whitespace after *
L:   3 | P:  53 | L010 | Keywords must be consistently upper case.
L:   4 | P:  21 | L011 | Implicit/explicit aliasing of table.
L:   4 | P:  21 | L031 | Avoid aliases in from clauses and join conditions.
L:   5 | P:  26 | L011 | Implicit/explicit aliasing of table.
L:   5 | P:  26 | L031 | Avoid aliases in from clauses and join conditions.
L:   6 | P:   1 | L003 | Indent expected and not found compared to line #5
==== fixing violations ====
15 fixable linting violations found
FORCE MODE: Attempting fixes...
Persisting Changes...
== [my_query.sql] PASS
Done. Please check your files to confirm.
```

`my_query.sql` was changed to:
```sql
SELECT
    abs(round(foo_bar_report.metricx - xxx_yyy_report.metricx)) AS col_c_rel_diff,
    abs((round(foo_bar_report.metricx - xxx_yyy_report.metricx) / foo_bar_report.metricx) * 100) AS metric_x_rel_diff
FROM foo_bar_report
LEFT JOIN xxx_yyy_report
    ON foo_bar_report.event_date = xxx_yyy_report.event_date;
```

Run fix again:
```
sqlfluff fix my_query.sql -f
```
It exits with 0.

Output:
```
==== finding fixable violations ====
== [my_query.sql] FAIL
L:   2 | P:  83 | L016 | Line is too long
L:   3 | P: 118 | L016 | Line is too long
==== fixing violations ====
2 fixable linting violations found
FORCE MODE: Attempting fixes...
Persisting Changes...
== [my_query.sql] PASS
Done. Please check your files to confirm.
```

`my_query.sql` was changed to:
```sql
SELECT
    abs(
        round(foo_bar_report.metricx - xxx_yyy_report.metricx)
    ) AS col_c_rel_diff,
    abs(
        (
            round(
                foo_bar_report.metricx - xxx_yyy_report.metricx
            ) / foo_bar_report.metricx
        ) * 100
    ) AS metric_x_rel_diff
FROM foo_bar_report
LEFT JOIN xxx_yyy_report
    ON foo_bar_report.event_date = xxx_yyy_report.event_date;
```

One more time:
```
sqlfluff fix my_query.sql -f
```
It exits with 0.

Output:
```
==== finding fixable violations ====
==== no fixable linting violations found ====
All Finished 📜 🎉!
```

Content of `my_query.sql` wasn't changed any more.

## Version

0.9.4

## Configuration

No custom conf

**Comment by juhoautio** [comment]
I didn't spot an answer to this question by @jpy-git (in https://github.com/sqlfluff/sqlfluff/issues/2134#issuecomment-997195047) yet:
> L006 and L031 are working exactly as they should, so the main question is why does L016 not just get applied at the next iteration of the fix loop?

That question sounds to me like it is not expected for sqlfluff to _not_ settle in this case. Could there be even some fundamental issue that is causing some cases like this to be missed? Or should this particular issue at least be fixed somehow?

**Comment by tunetheweb** [comment]
If you run `L031` first:

```
sqlfluff fix -f test.sql --rules L031
```

To get this SQL:

```
SELECT
    abs(round(foo_bar_report.metricx-xxx_yyy_report.metricx)) as col_c_rel_diff,
    abs((round(foo_bar_report.metricx-xxx_yyy_report.metricx)/foo_bar_report.metricx)*100) as metric_x_rel_diff
FROM foo_bar_report
LEFT JOIN xxx_yyy_report
ON foo_bar_report.event_date = xxx_yyy_report.event_date;
```

Then it corrects in one go.

This this (perhaps) suggests L031 is not replacing the alias's properly.

I do wonder again if we should re-parse the "fixed" SQL between iterations to avoid this as suggested in https://github.com/sqlfluff/sqlfluff/issues/1304#issuecomment-1030640659

**Comment by barrywhart** [comment]
Thanks for providing this case -- we'll look into it. In SQLFluff, each rule is independent of the other, and it's possible for them to interact in strange ways, including potentially "looping". Generally, we have fixed this by having a human take a look at both rules and figure out how to make them work well together. There haven't been many cases of this lately, but it could happen again as we add more rules or change existing ones.

Intellectually, fixing these issues at the individual rule level may seem unsatisfying -- you may think, "shouldn't there be some higher-level component that oversees what's happening to prevent all possible issues like this?" Maybe... if we knew how to do that, maybe we would. 😄

**Comment by juhoautio** [comment]
Thanks for yet another fix! Your speed of resolving bugs in this project is so impressive, I'm in awe.

Quoting @barrywhart's https://github.com/sqlfluff/sqlfluff/issues/2584#issuecomment-1033911820:

> Intellectually, fixing these issues at the individual rule level may seem unsatisfying -- you may think, "shouldn't there be some higher-level component that oversees what's happening to prevent all possible issues like this?" Maybe... if we knew how to do that, maybe we would.

I think there is a simple way to achieve that so that an error is raised if fix doesn't settle. It only requires this: 

_After fixing is complete, save the end result in a temporary cache. Then run fix once more on top of that result, and compare if anything would still be changed. If something **would** change, don't modify the actual files, but raise an error. If nothing would change any more, save the result (modify the actual files)._

Obviously this comes with some penalty in execution time. So I cannot tell if it's worth it, assuming that issues like this are rare.

Assuming that fix will settle _eventually_ it would be also totally possible to keep running the _entire_ fix process from "scratch" (keeping the end result of previous execution in cache), until the result settles. There could be some max number of rounds. For example in case of the issue here 2 rounds would have been enough. This option would be more user friendly in the short term, but on the other hand this would hide bugs instead of raising an error, so the bugs wouldn't be fixed and users would continue to suffer from the abnormally long execution time (if fix is internally executed many times instead of 1).

> There haven't been many cases of this lately, but it could happen again as we add more rules or change existing ones.

This is the tough part. I'm not familiar with the internals of sqlfluff, but it seems that at least the current design can't necessarily protect against new issues. But it's easy to see how an isolated check that happens outside of the whole fix process can guarantee either settling or raising an error. Should I create another issue to officially propose this top level check? I'm running out of reproducible case examples, but as you said "it could happen again as we add more rules or change existing ones" :)

**Comment by barrywhart** [comment]
I think we can do something like your suggested implementation can be implemented without much difficulty, because several components of it are already in place.

See the `previous_versions` variable [here](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/core/linter/linter.py#L468).

Then the linter loop adds the SQL from each "round" of linting [to a Python set](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/core/linter/linter.py#L525). This code was added to detect a previous type of bug, where the linter would get stuck in an endless cycle (e.g. two rules that alternate doing and undoing the same change).

Your suggestion sounds closely related to an existing issue. Can you take our discussion from this issue, summarize it, and post it as a comment to issue ##1386?

**Comment by juhoautio** [comment]
@barrywhart looks like you added a reference to my comment in there. Is that enough, or would you still like me to do this:

> Can you take our discussion from this issue, summarize it, and post it as a comment

**Comment by barrywhart** [comment]
@juhoautio: I think the reference I added is enough, but if there's anything you want to clarify, feel free to add an additional comment on #1386.
